### Scraping Instagram Data



### Use instagram-scraper to fetch the Instgram metadata associated with posts from these accounts. 

In [0]:
!pip install instagram-scraper 

  Created wheel for instagram-scraper: filename=instagram_scraper-1.8.1-cp36-none-any.whl size=34275 sha256=31c61352569e5a8530105047e3a52566114275af5a1b2971e38a420e841756b6
  Stored in directory: /root/.cache/pip/wheels/b6/53/4e/67f0a88142d60eac158eb38df3c7a159cc37358d638913a183
Successfully built instagram-scraper


In [0]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
!instagram-scraper ford,honda,volvocars,dodgeofficial,chevrolet --media-metadata -m 200 -d 'drive/My Drive/Colab Notebooks'

Searching ford for profile pic: 100% 1/1 [00:00<00:00, 1488.40 images/s]
Searching ford for posts: 151 media [00:28,  3.37s/ media]
Downloading: 100%|##########| 201/201 [00:02<00:00, 82.90it/s]
Searching honda for profile pic:   0% 0/1 [00:00<?, ? images/s]
Searching honda for profile pic: 100% 1/1 [00:00<00:00, 3412.78 images/s]
Searching honda for posts: 0 media [00:00, ? media/s]
Searching honda for posts: 1 media [00:15, 15.82s/ media]
Searching honda for posts: 51 media [00:30, 11.17s/ media]
Searching honda for posts: 101 media [00:45,  7.91s/ media]
Searching honda for posts: 151 media [01:02,  5.64s/ media]

Downloading:   0%|          | 0/201 [00:00<?, ?it/s]

Downloading:  76%|#######5  | 152/201 [00:00<00:00, 953.18it/s]

Downloading:  90%|######### | 181/201 [00:01<00:00, 68.02it/s] 

Downloading: 100%|##########| 201/201 [00:02<00:00, 68.99it/s]

Searching volvocars for profile pic:   0% 0/1 [00:00<?, ? images/s]

Searching volvocars for profile pic: 100% 1/1 [00:00<00:00

In [0]:
import glob
import json
import datetime
import csv
 
jsonfiles = glob.glob('drive/My Drive/Colab Notebooks/*.json')
 
for afile in jsonfiles:
    print(afile)
    outputfile = open('%s.csv' % (afile[:-5]), 'w')
     
    csvwriter = csv.writer(outputfile)
     
    jsonfile = open(afile, 'r')
    parsedjson = json.load(jsonfile)
    instagramposts = {}
     
    metadata = {}
    metadata['is_video'] = "Photo"
    metadata['display_url'] = ""
    metadata['caption'] = ""
    metadata['like count'] = ""
    metadata['comment count'] = ""
    metadata['tags'] = ""
    metadata['time'] = ""
    header = list(metadata.keys())
    csvwriter.writerows([header])
     
    for apost in parsedjson['GraphImages']:    
 
        metadata = {}
        metadata['is_video'] = "Photo"
        metadata['display_url'] = ""
        metadata['caption'] = ""
        metadata['like count'] = ""
        metadata['comment count'] = ""
        metadata['tags'] = ""
        metadata['time'] = ""
         
        if 'edge_media_to_caption' in apost:
            if 'edges' in apost['edge_media_to_caption']:
                edges = apost['edge_media_to_caption']['edges']
                if len(edges) > 0:
    #                    print("woo")
                    metadata['caption'] = apost['edge_media_to_caption']['edges'][0]['node']['text']
            if 'edge_media_preview_like' in apost:
                if 'count' in apost['edge_media_preview_like']:
                    metadata['like count'] = apost['edge_media_preview_like']['count']
            if 'edge_media_to_comment' in apost:
                if 'count' in apost['edge_media_to_comment']:
                    metadata['comment count'] = apost['edge_media_to_comment']['count']
            if 'tags' in apost:
                metadata['tags'] = apost['tags']
                     
            metadata['time'] = datetime.datetime.fromtimestamp(int(apost['taken_at_timestamp'])).strftime('%Y-%m-%d %H:%M')
            metadata['display_url'] = apost['display_url']
            if apost['is_video'] == True:
                metadata['is_video'] = "Video"
      
            row = []
            for anitem in metadata:
                row.append(metadata[anitem])
            csvwriter.writerow(row)
             
    outputfile.close()

drive/My Drive/Colab Notebooks/ford.json
drive/My Drive/Colab Notebooks/honda.json
drive/My Drive/Colab Notebooks/volvocars.json
drive/My Drive/Colab Notebooks/dodgeofficial.json
drive/My Drive/Colab Notebooks/chevrolet.json


###Calculate the average post engagement rate for each brand.


*I got the total numbers of followers by directly checking on the brand's Instagram profile page*

In [0]:
#total followers: 4.1 million 
import pandas as pd
Honda = pd.read_csv(open('drive/My Drive/Colab Notebooks/honda.csv','rb'))

Honda['post engagment rate']= (Honda['like count']+ Honda['comment count'])/4100000
Honda.head()


is_video  ... post engagment rate
0    Video  ...            0.000936
1    Photo  ...            0.003548
2    Video  ...            0.001252
3    Photo  ...            0.003165
4    Photo  ...            0.003141

[5 rows x 8 columns]

In [0]:
#Honda average post engagment rate
Honda_mean= Honda['post engagment rate'].mean()
Honda_mean

0.0042538414634146334

In [0]:
#total followers: 1.4 million 
import pandas as pd
volvo = pd.read_csv(open('drive/My Drive/Colab Notebooks/volvocars.csv','rb'))

volvo['post engagment rate']= (volvo['like count']+ volvo['comment count'])/1400000
volvo.head()

is_video  ... post engagment rate
0    Photo  ...            0.010961
1    Video  ...            0.002834
2    Photo  ...            0.021294
3    Photo  ...            0.009828
4    Photo  ...            0.015309

[5 rows x 8 columns]

In [0]:
#Volvo average post engagment rate: 
volvo_mean= volvo['post engagment rate'].mean()
volvo_mean

0.01381074285714286

In [0]:
#total followers: 5.3 million 
import pandas as pd
dodge = pd.read_csv(open('drive/My Drive/Colab Notebooks/dodgeofficial.csv','rb'))
dodge.head()

dodge['post engagment rate']= (dodge['like count']+ dodge['comment count'])/5300000
dodge.head()

is_video  ... post engagment rate
0    Photo  ...            0.019006
1    Video  ...            0.012826
2    Photo  ...            0.022331
3    Photo  ...            0.021832
4    Photo  ...            0.019292

[5 rows x 8 columns]

In [0]:
#dodgeofficia average post engagment rate:
dodge_mean= dodge['post engagment rate'].mean()
dodge_mean

0.021022040566037747

In [0]:
#total followers: 3 million
import pandas as pd
chevrolet = pd.read_csv(open('drive/My Drive/Colab Notebooks/chevrolet.csv','rb'))
chevrolet.head()

chevrolet['post engagment rate']= (chevrolet['like count']+ chevrolet['comment count'])/3000000
chevrolet.head()

is_video  ... post engagment rate
0    Photo  ...            0.008685
1    Photo  ...            0.017991
2    Photo  ...            0.013662
3    Photo  ...            0.018322
4    Video  ...            0.005301

[5 rows x 8 columns]

In [0]:
#chevrolet average post engagment rate:
chevrolet_mean= chevrolet['post engagment rate'].mean()
chevrolet_mean

0.01076851

In [0]:
#total followers: 3.9 million
import pandas as pd
ford = pd.read_csv(open('drive/My Drive/Colab Notebooks/ford.csv','rb'))
ford.head()

ford['post engagment rate']= (ford['like count']+ ford['comment count'])/3000000
ford.head()

is_video  ... post engagment rate
0    Photo  ...            0.013859
1    Photo  ...            0.033560
2    Photo  ...            0.049600
3    Photo  ...            0.013421
4    Photo  ...            0.014306

[5 rows x 8 columns]

In [0]:
#ford average post engagment rate: 
ford_mean= ford['post engagment rate'].mean()
ford_mean

0.017608106666666682

# Questions to answer

##1. Which of the brands has the highest average post engagement rate? 
  **Dodge** has the highest average post engagement rate which is around 2.1% 
##2. How is @honda's engagement rate compared to its competitors? 
**It has the lowest engagement rate (0.42%) compare to its competitors.**
